### PR TITLE
Enable Batch and Blas-like APIs for cuBLAS

### PR DESCRIPTION
**Pull request by mmeterel**
This PR is enabling Batch and Blas-like APIs for cuBLAS backend for the supported functions.

### CODE REVIEW

File: src/blas/backends/cublas/cublas_extensions.cpp [hunk 1]
```diff
@@ -73,44 +68,50 @@ void gemm_ext(cl::sycl::queue &queue, transpose transa, transpose transb, offset
...
-    throw backend_unsupported_exception();
-}
-
-void gemm_ext(cl::sycl::queue &queue, transpose transa, transpose transb, std::int64_t m,
-              std::int64_t n, std::int64_t k, double alpha, cl::sycl::buffer<double, 1> &a,
-              std::int64_t lda, cl::sycl::buffer<double, 1> &b, std::int64_t ldb, double beta,
-              cl::sycl::buffer<double, 1> &c, std::int64_t ldc) {
-    throw backend_unsupported_exception();
+template <typename Func, typename T_AB, typename T_C, typename DATATYPE_AB, typename DATATYPE_C>
+inline void gemm_ext(Func func, DATATYPE_AB DT_AB, DATATYPE_C DT_C, cl::sycl::queue &queue,
```
Comment (lhuot) [inline src/blas/backends/cublas/cublas_extensions.cpp:72]: To anticipate future functionality, could you use different template type for A and B matrices?

File: tests/unit_tests/blas/batch/gemm_batch_usm.cpp [hunk 2]
```diff
@@ -139,39 +134,53 @@ int test(const device &dev, int64_t group_count) {
...
+    c_array_data.resize(total_size_c);
+    c_ref_array_data.resize(total_size_c);
+
     idx = 0;
     for (i = 0; i < group_count; i++) {
         for (j = 0; j < group_size[i]; j++) {
-            a_array[idx] = (fp *)onemkl::malloc_shared(64, sizeof(fp) * total_size_a[i], dev, cxt);
-            b_array[idx] = (fp *)onemkl::malloc_shared(64, sizeof(fp) * total_size_b[i], dev, cxt);
-            c_array[idx] = (fp *)onemkl::malloc_shared(64, sizeof(fp) * total_size_c[i], dev, cxt);
-            c_ref_array[idx] =
```
Comment (lhuot) [inline tests/unit_tests/blas/batch/gemm_batch_usm.cpp:167]: Is there a reason why changing from independent pointers as allowed by the API to contiguous matrices?

File: tests/unit_tests/blas/batch/gemm_batch_usm.cpp [hunk 3]
```diff
@@ -139,39 +134,53 @@ int test(const device &dev, int64_t group_count) {
...
-            copy_matrix(c_array[idx], onemkl::transpose::nontrans, m[i], n[i], ldc[i],
-                        c_ref_array[idx]);
+            vector<fp, decltype(uafp)> a_array_temp(uafp), b_array_temp(uafp), c_array_temp(uafp),
+                c_ref_array_temp(uafp);
+            rand_matrix(a_array_temp, transa[i], m[i], k[i], lda[i]);
+            rand_matrix(b_array_temp, transb[i], k[i], n[i], ldb[i]);
+            rand_matrix(c_array_temp, onemkl::transpose::nontrans, m[i], n[i], ldc[i]);
+            copy_matrix(c_array_temp, onemkl::transpose::nontrans, m[i], n[i], ldc[i],
+                        c_ref_array_temp);
+            std::copy(a_array_temp.begin(), a_array_temp.end(), a_array_data.begin() + stride_a);
```
Comment (lhuot) [inline tests/unit_tests/blas/batch/gemm_batch_usm.cpp:172]: Those stride_? variable name are misleading, I would rather name them offset_?

File: tests/unit_tests/blas/batch/gemm_batch_usm.cpp (hunk 2 above)
Comment (mmeterel) [reply tests/unit_tests/blas/batch/gemm_batch_usm.cpp:167]: cuBLAS group API is expecting contiguous matrices. This change is done to support all backends.

File: src/blas/backends/cublas/cublas_extensions.cpp (hunk 1 above)
Comment (mmeterel) [reply src/blas/backends/cublas/cublas_extensions.cpp:72]: Makes sense. Will update.

File: tests/unit_tests/blas/batch/gemm_batch_usm.cpp (hunk 3 above)
Comment (mmeterel) [reply tests/unit_tests/blas/batch/gemm_batch_usm.cpp:172]: Sounds good.

File: src/blas/backends/cublas/cublas_extensions.cpp (hunk 1 above)
Comment (mmeterel) [reply src/blas/backends/cublas/cublas_extensions.cpp]: Template parameters are updated.

File: tests/unit_tests/blas/batch/gemm_batch_usm.cpp (hunk 3 above)
Comment (mmeterel) [reply tests/unit_tests/blas/batch/gemm_batch_usm.cpp]: Variable names are updated.

File: tests/unit_tests/blas/batch/gemm_batch_usm.cpp (hunk 2 above)
Comment (lhuot) [reply tests/unit_tests/blas/batch/gemm_batch_usm.cpp:167]: Do you have a pointer to the cuBLAS documentation that mentions this requirement? I could not find it.

Comment (mmeterel) [reply tests/unit_tests/blas/batch/gemm_batch_usm.cpp:167]: I take it back, you are right. I got confused. cuBLAS is also accepting array of pointers. 
Regarding contiguous matrices: Using malloc_shared for pointer arrays caused "CUDA_ERROR_ILLEGAL_ADDRESS" error in CUDA PI. As a workaround, I switched allocating array pointers and matrices using usm_allocator<T, usm::alloc::shared>. With that, I couldn't find a way to allocate matrices in a loop and still keep them alive beyond scope. Instead, I allocated a large array and copied the temporary matrices into this large array. This large array ended up being contiguous. In backends though, each array is accessed through pointers, so, they don't really care (or use) if the large array is contiguous or not.

File: src/blas/backends/cublas/cublas_batch.cpp [hunk 4]
```diff
@@ -96,78 +105,105 @@ void trsm_batch(cl::sycl::queue &queue, side left_right, uplo upper_lower, trans
...
+        cgh.interop_task([=](cl::sycl::interop_handler ih) {
+            auto sc        = CublasScopedContextHandler(queue);
+            auto handle    = sc.get_handle(queue);
+            int64_t stride = 0;
+            cublasStatus_t err;
+            for (int64_t i = 0; i < group_count; i++) {
+                auto **a_ = reinterpret_cast<const cuDataType **>(a);
+                auto **b_ = reinterpret_cast<const cuDataType **>(b);
+                auto **c_ = reinterpret_cast<cuDataType **>(c);
+                stride    = (i == 0) ? 0 : stride + group_size[i - 1];
```
Comment (lhuot) [inline src/blas/backends/cublas/cublas_batch.cpp:179]: offset for the variable name would be more appropriate maybe.
Also you can move it after the cublas call and remove the if statement using offset += group_size[i]

Comment (mmeterel) [reply src/blas/backends/cublas/cublas_batch.cpp:179]: Sounds good.

Comment (mmeterel) [reply src/blas/backends/cublas/cublas_batch.cpp]: Updated

File: tests/unit_tests/blas/batch/gemm_batch_usm.cpp (hunk 2 above)
Comment (mmeterel) [reply tests/unit_tests/blas/batch/gemm_batch_usm.cpp:167]: Updated memory allocation for matrices using vector of vectors.

Comment (mmeterel) [reply tests/unit_tests/blas/batch/gemm_batch_usm.cpp:167]: @lhuot Does this look good?

Comment (mmeterel) [reply tests/unit_tests/blas/batch/gemm_batch_usm.cpp:167]: After some study, I realized that, vector of vectors will also provide contiguous data array. So, I did more debug on why malloc_shared does not work in this case. The root-cause is using non-aligned memory. After fixing it, malloc_shared works fine and the non-contiguous data arrays are provided to the library calls.

File: tests/unit_tests/blas/batch/gemm_batch_usm.cpp [hunk 5]
```diff
@@ -116,40 +122,32 @@ int test(const device &dev, int64_t group_count) {
...
-    for (i = 0; i < group_count; i++) {
-        for (j = 0; j < group_size[i]; j++) {
-            a_array.push_back(&a_array_data[idx][0]);
-            b_array.push_back(&b_array_data[idx][0]);
-            c_array.push_back(&c_array_data[idx][0]);
-            c_ref_array.push_back(&c_ref_array_data[idx][0]);
+            a_array[idx]     = (fp *)onemkl::malloc_shared(64, sizeof(fp) * size_a[i], dev, cxt);
+            b_array[idx]     = (fp *)onemkl::malloc_shared(64, sizeof(fp) * size_b[i], dev, cxt);
+            c_array[idx]     = (fp *)onemkl::malloc_shared(64, sizeof(fp) * size_c[i], dev, cxt);
+            c_ref_array[idx] = (fp *)onemkl::malloc_shared(64, sizeof(fp) * size_c[i], dev, cxt);
```
Comment (lhuot) [inline tests/unit_tests/blas/batch/gemm_batch_usm.cpp:145]: size_{a,b,c} are used only here, I would suggest to not use array to store those sizes and move the size_{a,b,c} from the first loop of the test to the outer loop here.

File: tests/unit_tests/blas/batch/gemm_batch_usm.cpp (hunk 2 above)
Comment (lhuot) [reply tests/unit_tests/blas/batch/gemm_batch_usm.cpp:167]: Thanks!

File: tests/unit_tests/blas/batch/gemm_batch_usm.cpp (hunk 5 above)
Comment (mmeterel) [reply tests/unit_tests/blas/batch/gemm_batch_usm.cpp]: Refactored size_? parameters

**Pull request merged**